### DISCUSSION TITLE
generating-sortable-guids-using-newid/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#88] [Announcements]
# generating-sortable-guids-using-newid/

In this post I discuss the pros-and-cons of using integer vs Guids for IDs, and introduce NewId as an alternative to Guid with some nice properties

https://andrewlock.net/generating-sortable-guids-using-newid/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 03 2022, 05:34, in reply to [kapsiR](https://disqus.com/by/disqus_OMJgkwd063/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/88#discussioncomment-6063088)</em>

---

<p>Thanks! It's interesting to see that the level of fragmentation seen in this post caused up to a 32% impact in query duration, though that was for reading 100k rows, which can't be very common.</p><p>Another way to look at it is that NewId gives you a free query perf boost 😃</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 03 2022, 05:30, in reply to [Thomas Levesque](https://disqus.com/by/thomaslevesque/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/88#discussioncomment-6063093)</em>

---

<p>&gt;  If they're directly exposed in your API, it can reveal information about the data in your system<br>That's true. Is it a big issue do you think? It's not something I've every worried about, and if you want to use a large number for perception/vanity, you can start the primary keys at 10,000 😉 I guess NewId would largely hide that too.</p><p>&gt; Another issue is that if there are implementation flaws in the authorization system, a malicious user could possibly access data they shouldn't be able to access, just by incrementing the id</p><p>I thought about mentioning that, and decided not to on the basis that just because even though you can't get to it by incrementing by one, you still have an insecure direct object reference vulnerability when you use a guid 😃 I know <i>you're</i> aware of that, but part of me feels like hiding it behind a guid obscures the fact that there's an issue there. I wonder whether guids vs ints <i>actually</i> reduce how often an IDOR is exploited 🤔</p><p>&gt; One approach I like is to expose encrypted ids in the API, but it's quite cumbersome to implement. <br>Interesting, do you choose that over Guids generally?</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 05 2022, 08:24, in reply to [Jeremiah Peschka](https://disqus.com/by/jeremiahpeschka/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/88#discussioncomment-6063107)</em>

---

<p>&gt; If you want the GUIDs to actually sort over long periods of time you'll need to rearrange the byte order to match what SQL Server expects</p><p>Pretty sure that's what NewId is doing, but I haven't checked all the bit fiddling to be sure 🙂</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 05 2022, 08:28, in reply to [Scott](https://disqus.com/by/disqus_m27yWFKSCX/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/88#discussioncomment-6063111)</em>

---

<p>&gt; all that's happening here</p><p>That's a bit dismissive 😄 It's using those three components, yes, but taking care of storing them in suck a way that is efficient for SQL Server <i>and</i> is easy to use in code. In code, I have a single Guid I pass around which represents the entity ID. Having to pass around 3 fields is much less ergonomic.</p><p>&gt; without having to do horribly inefficient "guid" value processing<br>Where is the "horribly inefficient" guid processing? There's some bit fiddling in the c#, but it's primarily stack-based bit fiddling AFAICT, so no issues there, and I don't think SQL server does anything special with it?</p>

[a comment by an automated account is not shown]
